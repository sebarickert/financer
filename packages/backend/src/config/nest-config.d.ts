import { configuration } from './configuration';

import { FlattenConfig } from '@/types/flatten-config';

type ConfigurationType = FlattenConfig<
  Awaited<ReturnType<typeof configuration>>
>;

// Override NestJS configuration type and add types for configuration keys.
declare module '@nestjs/config' {
  class ConfigService {
    get<K extends keyof ConfigurationType>(
      key: K,
    ): K extends keyof ConfigurationType ? ConfigurationType[K] : never;
  }
}
