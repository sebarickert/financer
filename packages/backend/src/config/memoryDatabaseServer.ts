import { execSync } from 'child_process';

import { MongoMemoryReplSet } from 'mongodb-memory-server';

// Extend the default timeout so MongoDB binaries can download
// jest.setTimeout(60000);

const replSet = new MongoMemoryReplSet({
  replSet: { storageEngine: 'wiredTiger' },
  // configSettings: {
  //   getLastErrorDefaults: { w: 'majority', wtimeout: 5000 },
  // },
});

export const stopMemoryDb = async (): Promise<void> => {
  console.log('Stopping memory db');
  await replSet.stop();
};

export const getMemoryDbUri = async (): Promise<string> => {
  return replSet.getUri('financer_test');
};

export const startMemoryDb = async (): Promise<void> => {
  console.log('Starting memory db');
  await replSet.start();
  const uri = await getMemoryDbUri();

  console.log('Running prisma db push', uri);

  execSync(`npx prisma db push --schema=$SCHEMA --skip-generate`, {
    stdio: 'inherit',
    env: { ...process.env, DATABASE_URL: uri },
  });
};
