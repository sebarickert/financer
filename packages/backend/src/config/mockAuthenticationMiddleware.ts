import { Role, User } from '@prisma/client';
import { NextFunction, Request, Response } from 'express';

export const DUMMY_TEST_USER: Partial<User> = {
  roles: [Role.TEST_USER],
  id: '61460d7354ea082ad0256749',
  name: 'Dummy Test User',
};

export const mockAuthenticationMiddleware = (
  req: Request,
  _: Response,
  next: NextFunction,
): void => {
  req.user = DUMMY_TEST_USER;
  req.isAuthenticated = () => true;
  next();
};
