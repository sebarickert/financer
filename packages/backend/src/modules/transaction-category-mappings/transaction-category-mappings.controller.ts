import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
} from '@nestjs/common';

import { ObjectId } from '../../types/objectId';
import { LoggedIn } from '../auth/decorators/loggedIn.decorators';
import { UserId } from '../users/users.decorators';

import { CreateTransactionCategoryMappingDto } from './dto/create-transaction-category-mapping.dto';
import { UpdateTransactionCategoryMappingDto } from './dto/update-transaction-category-mapping.dto';
import { TransactionCategoryMappingsService } from './transaction-category-mappings.service';

@Controller('api/transaction-category-mappings')
@LoggedIn()
export class TransactionCategoryMappingsController {
  constructor(
    private readonly transactionCategoryMappingsService: TransactionCategoryMappingsService,
  ) {}

  @Post()
  create(
    @Body()
    createTransactionCategoryMappingDto: CreateTransactionCategoryMappingDto,
  ) {
    return this.transactionCategoryMappingsService.create(
      createTransactionCategoryMappingDto,
    );
  }

  @Get()
  findAllByUser(@UserId() userId: ObjectId) {
    return this.transactionCategoryMappingsService.findAllByUser(userId);
  }

  @Get(':id')
  findOne(@Param('id') id: string) {
    return this.transactionCategoryMappingsService.findOne(+id);
  }

  @Patch(':id')
  update(
    @Param('id') id: string,
    @Body()
    updateTransactionCategoryMappingDto: UpdateTransactionCategoryMappingDto,
  ) {
    return this.transactionCategoryMappingsService.update(
      +id,
      updateTransactionCategoryMappingDto,
    );
  }

  @Delete(':id')
  remove(@Param('id') id: string) {
    return this.transactionCategoryMappingsService.remove(+id);
  }
}
