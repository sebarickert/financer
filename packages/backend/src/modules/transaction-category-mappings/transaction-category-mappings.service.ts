import { BadRequestException, Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { TransactionCategoryMapping, TransactionType } from '@prisma/client';
import { Model } from 'mongoose';

import { TransactionCategoryMappingRepo } from '../../database/repos/transaction-category-mapping.repo';
import { ObjectId, parseObjectId } from '../../types/objectId';
import { CategoryMonthlySummaryDto } from '../transaction-categories/dto/transaction-month-summary.dto';

import { CreateTransactionCategoryMappingDto } from './dto/create-transaction-category-mapping.dto';
import {
  TransactionCategoryMapping as TransactionCategoryMappingOld,
  TransactionCategoryMappingDocument,
} from './schemas/transaction-category-mapping.schema';

@Injectable()
export class TransactionCategoryMappingsService {
  constructor(
    private readonly transactionCategoryMappingRepo: TransactionCategoryMappingRepo,
    @InjectModel(TransactionCategoryMappingOld.name)
    private transactionCategoryMappingModel: Model<TransactionCategoryMappingDocument>,
  ) {}

  async createMany(
    userId: string,
    createTransactionCategoryMappingDto: CreateTransactionCategoryMappingDto[],
  ) {
    return this.transactionCategoryMappingRepo.createMany(
      createTransactionCategoryMappingDto.map((category) => ({
        ...category,
        userId,
      })),
    );
  }

  findAll() {
    return `This action returns all transactionCategoryMappings`;
  }

  findOne(id: number) {
    return `This action returns a #${id} transactionCategoryMapping`;
  }

  async findAllByUser(userId: string): Promise<TransactionCategoryMapping[]> {
    return this.transactionCategoryMappingRepo.findMany({ where: { userId } });
  }

  async findAllByUserAndCategoryIds(
    userId: string,
    categoryIds: string[],
  ): Promise<TransactionCategoryMapping[]> {
    return this.transactionCategoryMappingRepo.findMany({
      where: {
        userId,
        categoryId: {
          in: categoryIds,
        },
      },
    });
  }

  async findAllByUserAndTransaction(
    userId: string,
    transactionId: string,
  ): Promise<TransactionCategoryMapping[]> {
    return this.transactionCategoryMappingRepo.findMany({
      where: { userId, transactionId },
    });
  }

  async findMonthlySummariesByUserAndId(
    userIdStr: string,
    categoryIds: string[],
    limit?: number,
    year?: number,
    month?: number,
  ): Promise<CategoryMonthlySummaryDto[]> {
    const userId = parseObjectId(userIdStr);

    return this.transactionCategoryMappingModel
      .aggregate([
        {
          $match: {
            owner: userId,
            category_id: { $in: categoryIds },
            ...this.getYearAndMonthFilter(year, month, 'laterThan'),
          },
        },
        {
          $lookup: {
            from: 'transactions',
            localField: 'transaction_id',
            foreignField: '_id',
            as: 'transaction',
          },
        },
        {
          $project: {
            date: {
              $arrayElemAt: ['$transaction.date', 0],
            },
            amount: '$amount',
            transactionAmount: {
              $arrayElemAt: ['$transaction.amount', 0],
            },
            fromAccount: {
              $arrayElemAt: ['$transaction.fromAccount', 0],
            },
            toAccount: {
              $arrayElemAt: ['$transaction.toAccount', 0],
            },
          },
        },
        {
          $group: {
            _id: {
              year: { $year: '$date' },
              month: { $month: '$date' },
            },
            totalCount: {
              $sum: await this.getMonthlySummaryCondition(userId, 1, null),
            },
            incomeCount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                1,
                TransactionType.INCOME,
              ),
            },
            expenseCount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                1,
                TransactionType.EXPENSE,
              ),
            },
            transferCount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                1,
                TransactionType.TRANSFER,
              ),
            },
            totalAmount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                '$amount',
                null,
              ),
            },
            totalTransactionAmount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                '$transactionAmount',
                null,
              ),
            },
            incomeAmount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                '$amount',
                TransactionType.INCOME,
              ),
            },
            incomeTransactionAmount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                '$transactionAmount',
                TransactionType.INCOME,
              ),
            },
            expenseAmount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                '$amount',
                TransactionType.EXPENSE,
              ),
            },
            expenseTransactionAmount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                '$transactionAmount',
                TransactionType.EXPENSE,
              ),
            },
            transferAmount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                '$amount',
                TransactionType.TRANSFER,
              ),
            },
            transferTransactionAmount: {
              $sum: await this.getMonthlySummaryCondition(
                userId,
                '$transactionAmount',
                TransactionType.TRANSFER,
              ),
            },
          },
        },
        {
          $sort: {
            _id: -1,
          },
        },
        {
          $project: {
            'total.count': '$totalCount',
            'total.amount': '$totalAmount',
            'total.transactionAmount': '$totalTransactionAmount',
            'income.count': '$incomeCount',
            'income.amount': '$incomeAmount',
            'income.transactionAmount': '$incomeTransactionAmount',
            'expense.count': '$expenseCount',
            'expense.amount': '$expenseAmount',
            'expense.transactionAmount': '$expenseTransactionAmount',
            'transfer.count': '$transferCount',
            'transfer.amount': '$transferAmount',
            'transfer.transactionAmount': '$transferTransactionAmount',
          },
        },
      ])
      .limit(limit || 1000)
      .exec();
  }

  private getYearAndMonthFilter(
    year?: number,
    month?: number,
    filterMode: 'targetMonth' | 'laterThan' = 'targetMonth',
  ) {
    if (!year && month) {
      throw new BadRequestException('Year is required when month is provided');
    }

    if (!year && !month) {
      return {};
    }

    if (filterMode === 'laterThan') {
      return {
        date: {
          $gte: new Date(year, month - 1 || 0, 1),
        },
      };
    }

    return {
      date: {
        $gte: new Date(year, month - 1 || 0, 1),
        $lt: new Date(year, month || 12, 1),
      },
    };
  }

  private async getMonthlySummaryCondition(
    userId: ObjectId,
    operator: '$amount' | '$transactionAmount' | 1 | 0,
    transactionType: TransactionType,
  ) {
    const selectedQuery =
      this.getAggregationTransactionTypeFilter(transactionType);

    if (transactionType !== null || typeof operator !== 'string') {
      return { $cond: [{ $and: selectedQuery }, operator, 0] };
    }

    return {
      $cond: [
        { $and: [...selectedQuery, { $eq: ['$fromAccount', null] }] },
        operator,
        {
          $cond: [
            { $and: [...selectedQuery, { $eq: ['$toAccount', null] }] },
            { $multiply: [operator, -1] },
            0,
          ],
        },
      ],
    };
  }

  private getAggregationTransactionTypeFilter(
    transactionType: TransactionType,
  ): { [key in string]: unknown }[] {
    const isEmpty = (fieldName) => ({ $eq: [fieldName, undefined] });
    const isNotEmpty = (fieldName) => ({ $ne: [fieldName, undefined] });

    switch (transactionType) {
      case TransactionType.INCOME:
        return [isNotEmpty('$toAccount'), isEmpty('$fromAccount')];
      case TransactionType.EXPENSE:
        return [isNotEmpty('$fromAccount'), isEmpty('$toAccount')];
      case TransactionType.TRANSFER:
        return [isNotEmpty('$fromAccount'), isNotEmpty('$toAccount')];
      case null:
        return [];
      default:
        throw new Error(`Invalid transaction type: ${transactionType}`);
    }
  }

  update(id: number) {
    return `This action updates a #${id} transactionCategoryMapping`;
  }

  remove(id: number) {
    return `This action removes a #${id} transactionCategoryMapping`;
  }

  async removeAllByUserAndTransaction(userId: string, transactionId: string) {
    await this.transactionCategoryMappingRepo.deleteMany({
      userId,
      transactionId,
    });
  }

  async removeAllByUser(userId: string) {
    await this.transactionCategoryMappingRepo.deleteMany({
      userId,
    });
  }
}
