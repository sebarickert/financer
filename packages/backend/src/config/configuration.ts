import { getMemoryDbUri } from './memoryDatabaseServer';

const isNotEmptyString = (value: string) => value && value.length > 0;

export const isNodeEnvInTest = () => process.env.NODE_ENV === 'test';

export const isGithubAuthEnabled = () =>
  !isNodeEnvInTest() &&
  isNotEmptyString(process.env.GITHUB_CLIENT_ID) &&
  isNotEmptyString(process.env.GITHUB_CLIENT_SECRET);

export const isAuth0AuthEnabled = () =>
  !isNodeEnvInTest() &&
  isNotEmptyString(process.env.AUTH0_DOMAIN) &&
  isNotEmptyString(process.env.AUTH0_CLIENT_ID) &&
  isNotEmptyString(process.env.AUTH0_CLIENT_SECRET);

const parseMongoDbUri = async (): Promise<string> => {
  if (!isNodeEnvInTest()) {
    return `mongodb://${process.env.DB_USER}:${process.env.DB_PASSWORD}@${process.env.DB_HOST}/${process.env.DB_NAME}`;
  }
  return getMemoryDbUri();
};

export const configuration = async () => ({
  port: parseInt(process.env.PORT, 10) || 3000,
  publicUrl: process.env.PUBLIC_URL,
  cookieKey: process.env.COOKIE_KEY
    ? process.env.COOKIE_KEY
    : 'thisappisawesome',
  mongodbConnectionString: await parseMongoDbUri(),
  githubKeys: isGithubAuthEnabled()
    ? {
        clientID: process.env.GITHUB_CLIENT_ID,
        clientSecret: process.env.GITHUB_CLIENT_SECRET,
      }
    : undefined,
  auth0Keys: isAuth0AuthEnabled()
    ? {
        domain: process.env.AUTH0_DOMAIN,
        clientID: process.env.AUTH0_CLIENT_ID,
        clientSecret: process.env.AUTH0_CLIENT_SECRET,
      }
    : undefined,
});
