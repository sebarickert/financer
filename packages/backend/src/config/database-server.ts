import { execSync } from 'child_process';

import { Logger } from '@nestjs/common';
import { PrismaClient } from '@prisma/client';
import { Client } from 'pg';

import {
  shouldOnlyExportApiSpec,
  shouldUseInternalDevelopmentDockerDb,
} from './configuration';
import { DUMMY_TEST_USER } from './mockAuthenticationMiddleware';

import { DockerDatabase } from '@/utils/docker-database';

const DEVELOPMENT_DB_PORT = 5432;
const TEST_DB_PORT = 39425;

// eslint-disable-next-line @typescript-eslint/no-extraneous-class
export class DatabaseServer {
  private static readonly logger = new Logger(DatabaseServer.name);

  private static readonly USERNAME = 'admin';

  private static readonly PASSWORD = 'change-me';

  public static readonly startServer = async (): Promise<void> => {
    const name = this.getContainerName();

    if (DockerDatabase.isContainerRunning(name)) {
      this.logger.verbose('PostgreSQL Docker container is already running.');
      return;
    } else if (DockerDatabase.isContainerExists(name)) {
      this.logger.verbose('Starting existing PostgreSQL Docker container.');
      await DockerDatabase.startContainer(name);
    }

    await DockerDatabase.createContainer(
      name,
      this.USERNAME,
      this.PASSWORD,
      this.getPort(),
    );
  };

  public static readonly stopServer = (): void => {
    const name = this.getContainerName();

    if (DockerDatabase.isContainerExists(name)) {
      this.logger.log(`Stopping PostgreSQL Docker container ${name}`);
      DockerDatabase.removeContainer(name);
    }
  };

  public static readonly setupTestDatabase = async (
    uriInput?: string,
  ): Promise<string> => {
    const uri = uriInput ?? this.getDbUri();

    await this.ensureDatabaseExists(uri);
    this.setupDatabaseSchema(uri);
    await this.ensureTestUserExists(uri);

    return uri;
  };

  public static readonly setupDevelopmentDatabase =
    async (): Promise<string> => {
      const uri = this.getDbUri();

      await this.ensureDatabaseExists(uri);
      this.setupDatabaseSchema(uri);

      return uri;
    };

  private static readonly getIsDevelopment = (): boolean => {
    return shouldUseInternalDevelopmentDockerDb() || shouldOnlyExportApiSpec();
  };

  private static readonly getContainerName = (): string => {
    const isDevelopment = this.getIsDevelopment();

    return isDevelopment
      ? 'financer-development-postgres'
      : 'financer-test-postgres';
  };

  private static getDbUri(): string {
    const databaseName = this.getIsDevelopment()
      ? 'financer_dev'
      : `test_db_${process.pid}`;

    return `postgresql://${this.USERNAME}:${this.PASSWORD}@localhost:${this.getPort()}/${databaseName}`;
  }

  private static getPort(): number {
    return this.getIsDevelopment() ? DEVELOPMENT_DB_PORT : TEST_DB_PORT;
  }

  private static async ensureDatabaseExists(uriInput: string): Promise<void> {
    const uri = new URL(uriInput);
    const dbName = uri.pathname.split('/').pop();
    uri.pathname = '/postgres';

    const client = new Client({ connectionString: uri.toString() });
    await client.connect();

    const res = await client.query(
      `SELECT 1 FROM pg_database WHERE datname='${dbName}'`,
    );
    if (res.rowCount === 0) {
      await client.query(`CREATE DATABASE "${dbName}"`);
      this.logger.log(`Database ${dbName} created.`);
    }

    await client.end();
  }

  private static readonly ensureTestUserExists = async (
    dbUri: string,
  ): Promise<void> => {
    const prisma = new PrismaClient({ datasources: { db: { url: dbUri } } });

    await prisma.user.upsert({
      where: { id: DUMMY_TEST_USER.id },
      update: {},
      create: DUMMY_TEST_USER,
    });

    await prisma.$disconnect();
  };

  private static readonly setupDatabaseSchema = (rawUri: string): void => {
    execSync(`npx prisma db push --schema=$SCHEMA --skip-generate`, {
      // Uncomment to see command output
      // Stdio: 'inherit',
      env: { ...process.env, DATABASE_URL: rawUri },
    });
  };
}
