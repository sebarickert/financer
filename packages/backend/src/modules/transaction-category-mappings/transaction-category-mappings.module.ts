import { Module } from '@nestjs/common';

import { DatabaseModule } from '../../database/database.module';

import { TransactionCategoryMappingsService } from './transaction-category-mappings.service';

@Module({
  imports: [DatabaseModule],
  providers: [TransactionCategoryMappingsService],
  exports: [TransactionCategoryMappingsService],
})
export class TransactionCategoryMappingsModule {}
