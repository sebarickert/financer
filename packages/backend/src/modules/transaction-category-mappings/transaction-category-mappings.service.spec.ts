import { Test, TestingModule } from '@nestjs/testing';

import { createMockServiceProvider } from '../../../test/create-mock-service-provider';
import { DUMMY_TEST_USER } from '../../config/mockAuthenticationMiddleware';
import {
  transactionCategoryMappingRepoFindAllByUserId,
  transactionCategoryMappingRepoFindAllByUserIdTransactionData,
} from '../../database/repos/mocks/transaction-category-mapping-repo-mock';
import { TransactionCategoryMappingRepo } from '../../database/repos/transaction-category-mapping.repo';

import { TransactionCategoryMappingsService } from './transaction-category-mappings.service';

describe('TransactionCategoryMappingsService', () => {
  let service: TransactionCategoryMappingsService;
  let transactionCategoryMappingRepo: jest.Mocked<TransactionCategoryMappingRepo>;

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        TransactionCategoryMappingsService,
        createMockServiceProvider(TransactionCategoryMappingRepo),
      ],
    }).compile();

    service = module.get<TransactionCategoryMappingsService>(
      TransactionCategoryMappingsService,
    );
    transactionCategoryMappingRepo = module.get<
      jest.Mocked<TransactionCategoryMappingRepo>
    >(TransactionCategoryMappingRepo);
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should return all transactionCategoryMappings by user', async () => {
    jest
      .spyOn(transactionCategoryMappingRepo, 'findMany')
      .mockResolvedValue(transactionCategoryMappingRepoFindAllByUserId);

    const transactionCategoryMappings = await service.findAllByUser(
      DUMMY_TEST_USER.id,
    );

    expect(transactionCategoryMappingRepo.findMany).toHaveBeenCalledTimes(1);
    expect(transactionCategoryMappingRepo.findMany).toHaveBeenCalledWith({
      where: {
        userId: '61460d7354ea082ad0256749',
      },
    });

    expect(transactionCategoryMappings).toMatchSnapshot();
  });

  it('should return all transactionCategoryMappings by user and category ids', async () => {
    jest
      .spyOn(transactionCategoryMappingRepo, 'findMany')
      .mockResolvedValue(transactionCategoryMappingRepoFindAllByUserId);

    const transactionCategoryMappings =
      await service.findAllByUserAndCategoryIds(DUMMY_TEST_USER.id, [
        '623b58ada3deba9879422fbf',
      ]);

    expect(transactionCategoryMappingRepo.findMany).toHaveBeenCalledTimes(1);
    expect(transactionCategoryMappingRepo.findMany).toHaveBeenCalledWith({
      where: {
        userId: '61460d7354ea082ad0256749',
        categoryId: {
          in: ['623b58ada3deba9879422fbf'],
        },
      },
    });

    expect(transactionCategoryMappings).toMatchSnapshot();
  });

  it('should return all transactionCategoryMappings by user and transaction id', async () => {
    jest
      .spyOn(transactionCategoryMappingRepo, 'findMany')
      .mockResolvedValue(transactionCategoryMappingRepoFindAllByUserId);

    const transactionCategoryMappings =
      await service.findAllByUserAndTransaction(
        DUMMY_TEST_USER.id,
        '624befb66ba655edad8f824e',
      );

    expect(transactionCategoryMappingRepo.findMany).toHaveBeenCalledTimes(1);
    expect(transactionCategoryMappingRepo.findMany).toHaveBeenCalledWith({
      where: {
        userId: '61460d7354ea082ad0256749',
        transactionId: '624befb66ba655edad8f824e',
      },
    });

    expect(transactionCategoryMappings).toMatchSnapshot();
  });

  it('should return transaction category monthly summary by user and category ids', async () => {
    jest
      .spyOn(transactionCategoryMappingRepo, 'findMany')
      .mockResolvedValue(
        transactionCategoryMappingRepoFindAllByUserIdTransactionData,
      );

    const transactionCategoryMappings =
      await service.findMonthlySummariesByUserAndId(DUMMY_TEST_USER.id, [
        '623b58ada3deba9879422fbf',
      ]);

    expect(transactionCategoryMappingRepo.findMany).toHaveBeenCalledTimes(1);
    expect(transactionCategoryMappingRepo.findMany).toHaveBeenCalledWith({
      include: {
        transaction: {
          select: {
            date: true,
            fromAccount: true,
            toAccount: true,
          },
        },
      },
      where: {
        categoryId: {
          in: ['623b58ada3deba9879422fbf'],
        },
        userId: '61460d7354ea082ad0256749',
      },
    });

    expect(transactionCategoryMappings).toMatchSnapshot();
  });
});
