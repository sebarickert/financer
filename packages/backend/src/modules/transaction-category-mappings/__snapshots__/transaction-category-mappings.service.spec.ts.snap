// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionCategoryMappingsService should return all transactionCategoryMappings by user 1`] = `
[
  {
    "amount": 500,
    "categoryId": "623b58ada3deba9879422fbf",
    "description": "",
    "id": "637e234fbda8050a85b1b6b7",
    "transactionId": "624befb66ba655edad8f824e",
    "userId": "61460d7354ea082ad0256749",
    "v": 0,
  },
  {
    "amount": 1358.18,
    "categoryId": "623b6b91a3deba9879422fe1",
    "description": "",
    "id": "637e234fbda8050a85b1b6b8",
    "transactionId": "624befb66ba655edad8f824e",
    "userId": "61460d7354ea082ad0256749",
    "v": 0,
  },
  {
    "amount": 635.98,
    "categoryId": "623b58ada3deba9879422fbf",
    "description": "",
    "id": "637e2371bda8050a85b1b70b",
    "transactionId": "624befb96ba655edad8f82ad",
    "userId": "61460d7354ea082ad0256749",
    "v": 0,
  },
  {
    "amount": 700,
    "categoryId": "623b58ada3deba9879422fbf",
    "description": "",
    "id": "637e2389bda8050a85b1b760",
    "transactionId": "624bef946ba655edad8f7f3d",
    "userId": "61460d7354ea082ad0256749",
    "v": 0,
  },
  {
    "amount": 1095.58,
    "categoryId": "623b6b91a3deba9879422fe1",
    "description": "",
    "id": "637e2389bda8050a85b1b761",
    "transactionId": "624bef946ba655edad8f7f3d",
    "userId": "61460d7354ea082ad0256749",
    "v": 0,
  },
  {
    "amount": 1571.53,
    "categoryId": "623b58ada3deba9879422fbf",
    "description": "",
    "id": "637e239cbda8050a85b1b7b8",
    "transactionId": "624bef936ba655edad8f7f33",
    "userId": "61460d7354ea082ad0256749",
    "v": 0,
  },
]
`;

exports[`TransactionCategoryMappingsService should return all transactionCategoryMappings by user and category ids 1`] = `
[
  {
    "amount": 500,
    "categoryId": "623b58ada3deba9879422fbf",
    "description": "",
    "id": "637e234fbda8050a85b1b6b7",
    "transactionId": "624befb66ba655edad8f824e",
    "userId": "61460d7354ea082ad0256749",
    "v": 0,
  },
  {
    "amount": 635.98,
    "categoryId": "623b58ada3deba9879422fbf",
    "description": "",
    "id": "637e2371bda8050a85b1b70b",
    "transactionId": "624befb96ba655edad8f82ad",
    "userId": "61460d7354ea082ad0256749",
    "v": 0,
  },
  {
    "amount": 700,
    "categoryId": "623b58ada3deba9879422fbf",
    "description": "",
    "id": "637e2389bda8050a85b1b760",
    "transactionId": "624bef946ba655edad8f7f3d",
    "userId": "61460d7354ea082ad0256749",
    "v": 0,
  },
  {
    "amount": 1571.53,
    "categoryId": "623b58ada3deba9879422fbf",
    "description": "",
    "id": "637e239cbda8050a85b1b7b8",
    "transactionId": "624bef936ba655edad8f7f33",
    "userId": "61460d7354ea082ad0256749",
    "v": 0,
  },
]
`;

exports[`TransactionCategoryMappingsService should return all transactionCategoryMappings by user and category ids 2`] = `
[
  {
    "expense": {
      "amount": 1135.98,
      "count": 2,
      "transactionAmount": 2494.16,
    },
    "id": {
      "month": 1,
      "year": 2022,
    },
    "income": {
      "amount": 2271.5299999999997,
      "count": 2,
      "transactionAmount": 3367.1099999999997,
    },
    "total": {
      "amount": 1135.55,
      "count": 4,
      "transactionAmount": 872.9499999999998,
    },
    "transfer": {
      "amount": 0,
      "count": 0,
      "transactionAmount": 0,
    },
  },
]
`;

exports[`TransactionCategoryMappingsService should return all transactionCategoryMappings by user and transaction id 1`] = `
[
  {
    "amount": 500,
    "categoryId": "623b58ada3deba9879422fbf",
    "description": "",
    "id": "637e234fbda8050a85b1b6b7",
    "transactionId": "624befb66ba655edad8f824e",
    "userId": "61460d7354ea082ad0256749",
    "v": 0,
  },
  {
    "amount": 1358.18,
    "categoryId": "623b6b91a3deba9879422fe1",
    "description": "",
    "id": "637e234fbda8050a85b1b6b8",
    "transactionId": "624befb66ba655edad8f824e",
    "userId": "61460d7354ea082ad0256749",
    "v": 0,
  },
]
`;
